From 54fc2ba9ee21e691d6a544850825881c5f9ea846 Mon Sep 17 00:00:00 2001
From: Ryan Blue <blue@apache.org>
Date: Wed, 1 Apr 2015 18:12:38 -0700
Subject: [PATCH 039/115] CDK-974: Update Hive DDL when updating a table.

Kite sets the table DDL when it creates a table and should update the
DDL if the schema changes.

Contributed by Andrew Stevenson.
---
 .../java/org/kitesdk/data/spi/hive/HiveUtils.java  |    4 +++
 .../org/kitesdk/data/spi/hive/TestHiveUtils.java   |   28 ++++++++++++++++++++
 2 files changed, 32 insertions(+), 0 deletions(-)

diff --git a/kite-data/kite-data-hive/src/main/java/org/kitesdk/data/spi/hive/HiveUtils.java b/kite-data/kite-data-hive/src/main/java/org/kitesdk/data/spi/hive/HiveUtils.java
index 22a8dad..be2f7e5 100644
--- a/kite-data/kite-data-hive/src/main/java/org/kitesdk/data/spi/hive/HiveUtils.java
+++ b/kite-data/kite-data-hive/src/main/java/org/kitesdk/data/spi/hive/HiveUtils.java
@@ -349,6 +349,10 @@ class HiveUtils {
 
     // keep the custom properties up-to-date
     addPropertiesForDescriptor(table, descriptor);
+
+    // keep the table DDL up to-to-date with the Schema
+    table.getSd().setCols(
+        HiveSchemaConverter.convertSchema(descriptor.getSchema()));
   }
 
   static FileSystem fsForPath(Configuration conf, Path path) {
diff --git a/kite-data/kite-data-hive/src/test/java/org/kitesdk/data/spi/hive/TestHiveUtils.java b/kite-data/kite-data-hive/src/test/java/org/kitesdk/data/spi/hive/TestHiveUtils.java
index 31f375d..8db4810 100644
--- a/kite-data/kite-data-hive/src/test/java/org/kitesdk/data/spi/hive/TestHiveUtils.java
+++ b/kite-data/kite-data-hive/src/test/java/org/kitesdk/data/spi/hive/TestHiveUtils.java
@@ -16,9 +16,11 @@
 
 package org.kitesdk.data.spi.hive;
 
+import org.apache.avro.SchemaBuilder;
 import org.apache.hadoop.conf.Configuration;
 import org.apache.hadoop.hive.conf.HiveConf;
 import org.apache.hadoop.hive.metastore.api.Table;
+import org.junit.Assert;
 import org.junit.Test;
 import static org.junit.Assert.*;
 import org.kitesdk.data.CompressionType;
@@ -43,6 +45,32 @@ public class TestHiveUtils {
   }
 
   @Test
+  public void testUpdateChangesDDL() throws Exception {
+    DatasetDescriptor original = new DatasetDescriptor.Builder()
+        .schema(SchemaBuilder.record("Test").fields()
+            .requiredLong("id")
+            .requiredString("data")
+            .endRecord())
+        .build();
+    boolean external = false;
+    Table table = HiveUtils.tableForDescriptor("ns", "test", original, external);
+
+    DatasetDescriptor updated = new DatasetDescriptor.Builder()
+        .schema(SchemaBuilder.record("Test").fields()
+            .requiredLong("id")
+            .requiredString("data")
+            .nullableString("data2", "")
+            .endRecord())
+        .build();
+
+    HiveUtils.updateTableSchema(table, updated);
+
+    Assert.assertEquals("Should update the table DDL",
+        table.getSd().getCols(),
+        HiveSchemaConverter.convertSchema(updated.getSchema()));
+  }
+
+  @Test
   public void testRoundTripDescriptorWithCompressionType() throws Exception {
     String namespace = "test_ns";
     String name = "test_table";
-- 
1.7.0.4

